mxcubecore/HardwareObjects/MicroDiffractometer.py

```diff
@@ -0,0 +1,475 @@
+
+Example xml file:
+<object class = "MicroDiffractometer"
+  <username>MD2S</username>
+  <exporter_address>wid30bmd2s:9001</exporter_address>
+  <motors>
+    <device role="omega" hwrid="/udiff_omega"/>
+    <device role="phiy" hwrid="/udiff_phiy"/>
+    <device role="phiz" hwrid="/udiff_phiz"/>
+    <device role="sampx" hwrid="/udiff_sampx"/>
+    <device role="sampy" hwrid="/udiff_sampy"/>
+    <device role="kappa" hwrid="/udiff_kappa"/>
+    <device role="kappa_phi" hwrid="/udiff_kappaphi"/>
+    <device role="focus" hwrid="/udiff_phix"/>
+    <device role="front_light" hwrid="/udiff_frontlight_intensity"/>
+    <device role="back_light" hwrid="/udiff_backlight_intensity"/>
+    <device role="horizontal_alignment" hwrid="/udiff_phiy"/>
+    <device role="vertical_alignment" hwrid="/udiff_phiz"/>
+    <device role="horizontal_centring" hwrid="/udiff_sampx"/>
+    <device role="vertical_centring" hwrid="/udiff_sampy"/>
+  </motors>
+  <nstate_equipment>
+    <object role="fast_shutter" href="/udiff_fastshut"/>
+    <object role="scintillator" href="/udiff_scint"/>
+    <object role="diode" href="/udiff_diode"/>
+    <object role="fluo_detector" href="/udiff_fluodet"/>
+    <object role="cryostream" href="/udiff_cryostream"/>
+    <object role="front_light" href="/udiff_frontlight_inout"/>
+    <object role="back_light" href="/udiff_backlight_inout"/>
+    <object role="beamstop" href="/udiff_beamstop"/>
+    <object role="capillary" href="/udiff_capillary"/>
+    <object role="aperture" href="/udiff_aperturemot"/>
+    <object role="zoom" href="/udiff_zoom"/>
+  </nstate_equipment>
+</object>
+"""
+
+from gevent import Timeout, sleep
+
+from mxcubecore.Command.Exporter import Exporter
+from mxcubecore.Command.exporter.ExporterStates import ExporterStates
+from mxcubecore.HardwareObjects.abstract.AbstractDiffractometer import (
+    AbstractDiffractometer,
+    DiffractometerConstraint,
+    DiffractometerHead,
+    DiffractometerPhase,
+)
+
+__copyright__ = """ Copyright © by the MXCuBE collaboration """
+__license__ = "LGPLv3+"
+
+
+class MicroDiffractometer(AbstractDiffractometer):
+    """Microdiff with Exporter implementation of AbstartDiffractometer"""
+
+    def __init__(self, name):
+        super().__init__(name)
+        self._exporter = None
+
+    def init(self):
+        """Initialise the device"""
+        super().init()
+        # Initialise the commands and channels
+        exporter_address = self.get_property("exporter_address")
+        _host, _port = exporter_address.split(":")
+        self._exporter = Exporter(_host, int(_port))
+        self.head_type = self._get_head_type
+        self.update_state(self.get_state())
+
+    def abort(self):
+        """Immediately terminate action."""
+        self._exporter.execute("abort")
+
+    @property
+    def _get_hwstate(self) -> str:
+        """Get the hardware state, reported by the MD2 application.
+        Returns:
+            (str): The state.
```

Comment:
There has to be an empty line.

There is no need to have types in the docstring that are already documented as type hint annotations.

```suggestion
        """Get the hardware state, reported by the MD2 application.
        
        Returns:
            The state.
```

Reply:
> There has to be an empty line.

I was relying on the pre-commit to sort this kind of formatting.

Reply:
Sadly nearly all docstring-related rules are disabled globally currently:

<https://github.com/mxcube/mxcubecore/blob/d295958e141b51088967eb38c1af7c4cd1d4eb55/ruff.toml#L11>

I did this to make the transition easier. At the time I thought I would get around to improving this situation quickly after creating this situation, but that did not happen. I should come up with a plan how to get out of this inconvenient situation.